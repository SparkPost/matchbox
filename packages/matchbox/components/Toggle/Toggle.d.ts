export default Toggle;
declare function Toggle(props: any): any;
declare namespace Toggle {
    const displayName: string;
    const propTypes: any;
}
