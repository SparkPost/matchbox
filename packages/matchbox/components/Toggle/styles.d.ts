export function toggle(props: any): string;
export const StyledIndicator: any;
export const StyledOutline: any;
export function input(): string;
