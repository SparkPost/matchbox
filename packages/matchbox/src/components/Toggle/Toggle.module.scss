@import '../../styles/config';

.Toggle {
  display: inline-block;
  position: relative;
  height: rem(30);
  width: rem(60);

  *:hover {
    cursor: pointer;
  }
}

.compact {
  height: rem(21);
  width: rem(36);

  .Indicator {
    top: rem(3);
    left: rem(3);
    bottom: rem(3);
    width: rem(15);
  }

  .Input:checked ~ .Indicator {
    transform: translate(rem(15), 0);
  }

  .Input:focus ~ .Outline,
  .Input:active ~ .Outline {
    background: color(gray, 6);
  }

  .Outline {
    background: color(gray, 5);
  }

  .Input:checked ~ .Outline {
    background: color(green);
  }
}

.disabled {
  opacity: 0.9;

  *, *:hover {
    cursor: not-allowed;
  }

  .Labels {
    span {
      opacity: 0.4;
    }
  }
}

.Input {
  @include visually-hidden;

  &:focus ~ .Outline,
  &:active ~ .Outline {
    box-shadow: 0 0 0 1px white, 0 0 0 3px color(blue);
  }

  &:checked {
    & ~ .Indicator {
      transform: translate(rem(29), 0);
    }

    & ~ .Outline {
      background: color(green);
    }
  }
}

.Outline {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  background: color(gray, 7);
  border-radius: rem(15);
  border: 1px solid color(gray, 9);
  transition: 0.15s;
}

.Indicator {
  pointer-events: none;
  position: absolute;
  top: rem(3);
  left: rem(3);
  bottom: rem(3);
  width: rem(25);

  background: color(gray, 10);
  transition: 0.15s;

  border-radius: 50%;
  box-shadow: shadow(deep);
}

.Labels {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  user-select: none;

  span {
    position: absolute;
    top: 2px;
    font-size: font-size(200);
    font-weight: 500;
    line-height: 2em;

    &:first-child {
      left: rem(9);
      color: color(gray, 10);
    }

    &:last-child {
      right: rem(10);
    }
  }
}
