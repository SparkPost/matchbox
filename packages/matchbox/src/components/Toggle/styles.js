import { tokens } from '@sparkpost/design-tokens';
import styled from 'styled-components';

const toggleWidth = '2.25rem';

export const toggle = props => `
  display: inline-block;
  position: relative;
  height: 1.25rem;
  width: ${toggleWidth};

  opacity: ${props.disabled ? '0.6' : '1'};
  cursor: ${props.disabled ? 'not-allowed' : 'pointer'};
`;

export const StyledIndicator = styled('span')`
  pointer-events: none;
  position: absolute;
  top: 1px;
  left: 1px;
  bottom: 1px;
  width: calc(${toggleWidth} / 2);

  background: ${tokens.color_white};
  transition: transform ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
  transform: translate(${props => (props.checked ? tokens.spacing_400 : '0')}, 0);

  border-radius: ${tokens.borderRadius_circle};
  box-shadow: ${tokens.boxShadow_100};
`;

export const StyledOutline = styled('span')`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  background: ${tokens.color_gray_700};
  border-radius: ${tokens.borderRadius_pill};
  transition: ${tokens.motionDuration_fast};
`;

export const input = () => `
  &:focus ~ ${StyledOutline},
  &:active ~ ${StyledOutline} {
    box-shadow: 0 0 0 2px white, 0 0 0 4px ${tokens.color_blue_600};
  }

  &:checked {
    & ~ ${StyledIndicator} {
      transform: translate(${tokens.spacing_400}, 0);
    }

    & ~ ${StyledOutline} {
      background: ${tokens.color_green_800};
    }
  }
`;
