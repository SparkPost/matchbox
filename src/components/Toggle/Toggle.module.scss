@import '../../styles/config';

.Toggle {
  display: inline-block;
  position: relative;
  height: rem(30);
  width: rem(60);

  *:hover {
    cursor: pointer;
  }
}

.Input {
  @include visually-hidden;

  &:focus ~ .Outline,
  &:active ~ .Outline {
    box-shadow: 0 0 0 1px color(blue);
    border: 1px solid color(blue);
  }

  &:checked {
    & ~ .Indicator {
      transform: translate(rem(29), 0);
    }

    & ~ .Outline {
      background: color(blue);
    }
  }

  &:disabled {
    cursor: not-allowed;
    & ~ .Indicator {
      background: color(gray, 9);
      cursor: not-allowed;
    }

    & ~ .Labels {
      span {
        opacity: 0.8;
        cursor: not-allowed;
      }
    }
  }
}

.Outline {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  background: color(gray, 8);
  border-radius: rem(15);
  border: 1px solid color(gray, 9);
  transition: 0.15s;
}

.Indicator {
  pointer-events: none;
  position: absolute;
  top: rem(3);
  left: rem(3);
  bottom: rem(3);
  width: rem(25);

  background: color(gray, 10);
  transition: 0.15s;

  border-radius: 100%;
  box-shadow: shadow(deep);
}

.Labels {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  user-select: none;

  span {
    position: absolute;
    top: rem(3);
    font-size: rem(12);
    font-weight: 500;

    &:first-child {
      left: rem(9);
      color: color(gray, 10);
    }

    &:last-child {
      right: rem(10);
    }
  }
}
