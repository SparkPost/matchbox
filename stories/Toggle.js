import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { Toggle } from '../src';

class CheckboxWrapper extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      checked: props.checked
    }
  }

  handleChange() {
    this.setState({
      checked: !this.state.checked
    });
  }

  render() {
    const { label } = this.props;
    return (
      <Checkbox
        id={label}
        label={label}
        checked={this.state.checked}
        onChange={ () => this.handleChange() }
      />
    )
  }
};

export default storiesOf('Toggle', module)
  .addDecorator((getStory) => (
    <StoryContainer bg='white'>{ getStory() }</StoryContainer>
  ))

  .add('Default', withInfo()(() => (
    <Toggle
      id='id'
     />
  )))

  .add('Disabled', withInfo()(() => (
    <Toggle
      id='id'
      disabled
     />
  )));
