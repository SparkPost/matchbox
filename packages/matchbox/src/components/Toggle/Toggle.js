import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './Toggle.module.scss';

class Toggle extends Component {
  static displayName = 'Toggle';

  static propTypes = {
    id: PropTypes.string.isRequired,
    checked: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string
    ]),
    compact: PropTypes.bool,
    value: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string
    ]),
    disabled: PropTypes.bool,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func
  };

  render() {
    const {
      id,
      value,
      checked,
      compact,
      disabled,
      onChange,
      onFocus,
      onBlur,
      ...rest
    } = this.props;

    const labelMarkup = compact ? null : (
      <div className={styles.Labels}>
        <span>On</span>
        <span>Off</span>
      </div>
    );

    return (
      <label htmlFor={id} className={classnames(styles.Toggle, compact && styles.compact, disabled && styles.disabled)}>
        <input
          id={id}
          value={value}
          checked={checked}
          disabled={disabled}
          className={styles.Input}
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          type='checkbox'
          {...rest} />
        <span className={styles.Outline}></span>
        {labelMarkup}
        <span className={styles.Indicator}></span>
      </label>
    );
  }
}

export default Toggle;
