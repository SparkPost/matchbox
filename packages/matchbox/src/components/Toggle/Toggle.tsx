import React from 'react';
import styled from 'styled-components';
import { margin, MarginProps } from 'styled-system';
import { pick } from '@styled-system/props';
import { visuallyHidden } from '../../styles/helpers';
import { omit } from '../../helpers/props';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { toggle, input, StyledIndicator, StyledOutline } from './styles';

const StyledToggle = styled('label')`
  ${toggle}
  ${margin}
`;

const StyledInput = styled('input')`
  ${visuallyHidden}
  ${input}
`;

export type ToggleProps = React.ComponentPropsWithRef<'input'> &
  MarginProps & {
    label?: React.ReactNode;
    'data-id'?: string;
  };

function Toggle(props: ToggleProps): JSX.Element {
  const {
    id,
    value,
    checked,
    defaultChecked,
    disabled,
    onChange,
    onFocus,
    onBlur,
    label,
    required,
    'aria-describedby': ariaDescribedBy,
    'data-id': dataId,
    ...rest
  } = props;
  const systemProps = pick(rest);
  const componentProps = omit(rest, margin.propNames);

  return (
    <StyledToggle htmlFor={id} disabled={disabled} {...systemProps}>
      {label && <ScreenReaderOnly>{label}</ScreenReaderOnly>}
      <StyledInput
        id={id}
        value={value}
        defaultChecked={defaultChecked}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
        required={required}
        type="checkbox"
        aria-describedby={ariaDescribedBy}
        data-id={dataId}
        {...componentProps}
      />
      <StyledOutline />
      <StyledIndicator $checked={checked} />
    </StyledToggle>
  );
}

Toggle.displayName = 'Toggle';
export default Toggle;
