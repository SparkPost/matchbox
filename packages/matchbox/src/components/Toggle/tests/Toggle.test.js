import React from 'react';
import Toggle from '../Toggle';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

describe('Toggle', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Toggle id='foo' value='bar'/>);
  });

  const testCases = [
    { name: 'checked', props: { checked: true }},
    { name: 'disabled', props: { disabled: true }},
    { name: 'compact', props: { compact: true }}
  ];

  cases('renders toggle states', (opts) => {
    wrapper.setProps(opts.props);
    expect(wrapper).toMatchSnapshot();
  }, testCases);

  cases('invokes event', (opts) => {
    const fn = jest.fn();
    const newProps = {};
    newProps[opts.name] = fn;
    wrapper.setProps(newProps);

    wrapper.find('input').simulate(opts.event);
    expect(fn).toHaveBeenCalledTimes(1);
  }, [
    { name: 'onClick', event: 'click' },
    { name: 'onBlur', event: 'blur' },
    { name: 'onFocus', event: 'focus' }
  ]);
});
