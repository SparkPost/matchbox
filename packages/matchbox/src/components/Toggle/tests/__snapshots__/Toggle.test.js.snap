// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders toggle states checked 1`] = `
<label
  className="Toggle"
  htmlFor="foo"
>
  <input
    checked={true}
    className="Input"
    id="foo"
    type="checkbox"
    value="bar"
  />
  <span
    className="Outline"
  />
  <div
    className="Labels"
  >
    <span>
      On
    </span>
    <span>
      Off
    </span>
  </div>
  <span
    className="Indicator"
  />
</label>
`;

exports[`Toggle renders toggle states compact 1`] = `
<label
  className="Toggle compact"
  htmlFor="foo"
>
  <input
    className="Input"
    id="foo"
    type="checkbox"
    value="bar"
  />
  <span
    className="Outline"
  />
  <span
    className="Indicator"
  />
</label>
`;

exports[`Toggle renders toggle states disabled 1`] = `
<label
  className="Toggle disabled"
  htmlFor="foo"
>
  <input
    className="Input"
    disabled={true}
    id="foo"
    type="checkbox"
    value="bar"
  />
  <span
    className="Outline"
  />
  <div
    className="Labels"
  >
    <span>
      On
    </span>
    <span>
      Off
    </span>
  </div>
  <span
    className="Indicator"
  />
</label>
`;
